### PR TITLE
Add clearer exception for read failures in macro plugin upgrade check

### LINKED ISSUE #13349: AirflowMacroPluginRemovedRule fails on non-python files
**Apache Airflow version**: 1.10.14

**Kubernetes version (if you are using kubernetes)** (use `kubectl version`):

**Environment**:

- **Cloud provider or hardware configuration**: X
- **OS** (e.g. from /etc/os-release): X
- **Kernel** (e.g. `uname -a`): X
- **Install tools**: X
- **Others**: X

**What happened**:

The `AirflowMacroPluginRemovedRule` seems unable to process non-standard python files (e.g. `.xlsx`) and chokes out with an unhelpful error message.:

```python
========================================================================================================================================================== STATUS ==========================================================================================================================================================

Check for latest versions of apache-airflow and checker...........................................................................................................................................................................................................................................................SUCCESS
Traceback (most recent call last):
  File "/Users/madison/programs/anaconda3/envs/memphis-airflow/bin/airflow", line 37, in <module>
    args.func(args)
  File "/Users/madison/programs/anaconda3/envs/memphis-airflow/lib/python3.8/site-packages/airflow/upgrade/checker.py", line 88, in run
    all_problems = check_upgrade(formatter, rules)
  File "/Users/madison/programs/anaconda3/envs/memphis-airflow/lib/python3.8/site-packages/airflow/upgrade/checker.py", line 37, in check_upgrade
    rule_status = RuleStatus.from_rule(rule)
  File "/Users/madison/programs/anaconda3/envs/memphis-airflow/lib/python3.8/site-packages/airflow/upgrade/problem.py", line 44, in from_rule
    result = rule.check()
  File "/Users/madison/programs/anaconda3/envs/memphis-airflow/lib/python3.8/site-packages/airflow/upgrade/rules/airflow_macro_plugin_removed.py", line 52, in check
    problems.extend(self._check_file(file_path))
  File "/Users/madison/programs/anaconda3/envs/memphis-airflow/lib/python3.8/site-packages/airflow/upgrade/rules/airflow_macro_plugin_removed.py", line 42, in _check_file
    for line_number, line in enumerate(file_pointer, 1):
  File "/Users/madison/programs/anaconda3/envs/memphis-airflow/lib/python3.8/codecs.py", line 322, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x82 in position 16: invalid start byte
```

**What you expected to happen**:

I expected the macro to skip over files it could not process/understand

**How to reproduce it**:

Add an `.xlsx` or other binary document to the DAGs folder and run the upgrade check.

**Suggested resolution**:

I think it's fine to fail out on these files (it led us to add certain items to the `.airflowignore` which should have been there anyway) but I had to modify the upgrade rule directly to tell me _which_ files were the problem. A more helpful error message here, and possibly a message prompting users to add said files to their `.airflowignore` would be ideal.

### PR DESCRIPTION
**Pull request by madison-ookla**
Resolves #13349

This PR adds a more useful error message when processing files using the `AirflowMacroPluginRemovedRule` upgrade check. It also prompts users to add said files to the `.airflowignore` file (which resolved the issue in our case).

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Comment by madison-ookla** [comment]
I'm hoping I've done everything correctly...apologies if not! It's my first time contributing to airflow in quite some time 😄

File: airflow/upgrade/rules/airflow_macro_plugin_removed.py
Comment (potiuk) [inline airflow/upgrade/rules/airflow_macro_plugin_removed.py]: I think it would also be great to add the original exception as a "chained" exception https://docs.python.org/3/tutorial/errors.html#exception-chaining this way we can see also the original stack trace.

File: airflow/upgrade/rules/airflow_macro_plugin_removed.py
Comment (potiuk) [inline airflow/upgrade/rules/airflow_macro_plugin_removed.py]: I think it would be much better to fix the problem as well for non-default encodings.

As of python 3.4 there is this function in the importlib:

https://docs.python.org/3/library/importlib.html?highlight=importlib#importlib.util.decode_source

And it will read the source code of python, respecting the encoding defined in the source code( the `# coding = ` comment). This way it will be more universal and handle the cases where non-utf8 encoding is used (which might be important in many countries). 

I am not sure if similar approach is used in other rules - if so, the we can make a common approach and correct it in all those places.

**Review by potiuk — CHANGES_REQUESTED** [review]

File: airflow/upgrade/rules/airflow_macro_plugin_removed.py
Comment (madison-ookla) [reply airflow/upgrade/rules/airflow_macro_plugin_removed.py]: I actually added this initially! But because this is against the 1.10 stable branch, it failed on the python2 check 😞

File: airflow/upgrade/rules/airflow_macro_plugin_removed.py
Comment (madison-ookla) [reply airflow/upgrade/rules/airflow_macro_plugin_removed.py]: I think this might run into the same [python 2.7](https://docs.python.org/2/library/importlib.html) issue as the [other comment you made](https://github.com/apache/airflow/pull/13371/files#r550270606) 😮  I'm not opposed to changing this logic, if you have an alternate suggestion! I think the issue I was running into had to deal with the fact that the file was binary, not that it was in a different encoding (as I imagine `open` would handle other string encodings appropriately...but admittedly I know less about that side of things).

Comment (potiuk) [reply airflow/upgrade/rules/airflow_macro_plugin_removed.py]: This is no problem. Just skip the test for py < 3.4. The upgradecheck is supposed to be run after people migrate to Python 3.6+. Something like that (or using six) should do.

```
@pytest.mark.skipif(sys.version_info < (3, 4), reason="requires python3.4 or higher")
```

File: airflow/upgrade/rules/airflow_macro_plugin_removed.py
Comment (potiuk) [reply airflow/upgrade/rules/airflow_macro_plugin_removed.py]: SkipIf :) - see below.

File: airflow/upgrade/rules/airflow_macro_plugin_removed.py
Comment (potiuk) [reply airflow/upgrade/rules/airflow_macro_plugin_removed.py]: I think we should be completely safe to put skipIf < 3.6 on all upgrade check tests @turbaszek @dimberman ?

Comment (madison-ookla) [reply airflow/upgrade/rules/airflow_macro_plugin_removed.py]: Gotcha! This was caught by the `Compile code using python2` pre-commit hook (run via breeze), which is what clued me in initially.

Comment (potiuk) [reply airflow/upgrade/rules/airflow_macro_plugin_removed.py]: Ah ok. Then if it a that stage, that might be a bit more complex, but just a bit. Simply exclude upgrade_check tests + upgrade_check sources in those two pre-commit checks (they already contain some exclusions):  https://github.com/apache/airflow/blob/748d05f519c0e5427d2a46763591f03697c4d585/.pre-commit-config.yaml#L266

Comment (madison-ookla) [reply airflow/upgrade/rules/airflow_macro_plugin_removed.py]: Can do! I'll wait until some others chime in on whether or not this can be 3.6+ exclusively, then get to work on that if we get the 👍  🙂

Comment (turbaszek) [inline airflow/upgrade/rules/airflow_macro_plugin_removed.py]: Imho we should add this info to `problem` instead of rising the exception. In this way the command will work and users will be informed about all problems - including undesired files. WDYT?

Btw. shouldn't we only check files with `.py` extension? Currently we open every file and I as far as I'm aware macros have to be python files not, `xlsx` for example 😄

Comment (madison-ookla) [reply airflow/upgrade/rules/airflow_macro_plugin_removed.py]: More than happy to add this to the problems report! My gut instinct for big projects like this is "preserve existing behavior by default", so I didn't want to change things up too much initially 😄  Also good to know on the macros! I can add that to the logic as well. Can they be python files that don't have the `.py` extension but are ostensibly python otherwise?

Comment (turbaszek) [reply airflow/upgrade/rules/airflow_macro_plugin_removed.py]: Plugins can only be python files with `.py` extension, otherwise we skip them:
https://github.com/apache/airflow/blob/748d05f519c0e5427d2a46763591f03697c4d585/airflow/plugins_manager.py#L200-L201

Comment (madison-ookla) [reply airflow/upgrade/rules/airflow_macro_plugin_removed.py]: Just want to check back in on this @potiuk @turbaszek @dimberman - we can make the upgrade check Python 3.6+ (or even just the tests Py3.6+ if we'd like), but currently the package still supports 2.7 (https://github.com/apache/airflow/blob/v1-10-stable/airflow/upgrade/setup.cfg#L35). Should we bump up these requirements to 3.6? I agree that it makes sense to do so, since this is a utility for upgrading to Airflow 2.0 which is 3.6+.

Comment (turbaszek) [reply airflow/upgrade/rules/airflow_macro_plugin_removed.py]: The `airflow upgrade_check` is intended to run on `1.10.X` version of Airflow so I'm afraid that it has to support Py2.7

Comment (madison-ookla) [reply airflow/upgrade/rules/airflow_macro_plugin_removed.py]: Thanks @turbaszek! With that in mind @potiuk, I'll make sure this is python 2.7 compatible. I'm going to be changing it up a bit though to raise these issues as problems rather than an exception, per the comments below.

**Comment by madison-ookla** [comment]
Updated to retain support for python 2, only check python files, and accumulate exceptions as problems rather than re-raising exceptions!

**Review by turbaszek — APPROVED** [review]
LGTM 🚀

**Comment by madison-ookla** [comment]
@potiuk Can this be merged?

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
@madison-ookla  Can you please rebase (or commit --amend followed by --force-push ? ) I think there were some CI woes when you pushed it last time and we did not have tests run for this one.

**Comment by madison-ookla** [comment]
> @madison-ookla Can you please rebase (or commit --amend followed by --force-push ? ) I think there were some CI woes when you pushed it last time and we did not have tests run for this one.

Done!

**Comment by madison-ookla** [comment]
Is this ready to be merged?

**Comment by kaxil** [comment]
Unfortunately, the tests were not running on v1-10-stable -- I just fixed it and rebased your PR -- 🤞 let's wait for CI results before I merge this one

**Comment by madison-ookla** [comment]
It might be my lack of understanding, but it seems the CI failures are unrelated?

**Review by kaxil — APPROVED** [review]
Yup, passed:

```
tests/upgrade/rules/test_airflow_macro_plugin_removed.py .....           [ 85%]

```

**Comment by kaxil** [comment]
@madison-ookla This PR failed tests for Python 2.7 -- can you take a look please and fix it if possible in a separate PR -- thanks :)

**Comment by madison-ookla** [comment]
> @madison-ookla This PR failed tests for Python 2.7 -- can you take a look please and fix it if possible in a separate PR -- thanks :)
> 
> https://github.com/apache/airflow/runs/1770939841#step:6:821

Oh dear....yes, I'll look into this!

**Comment by kaxil** [comment]
> > @madison-ookla This PR failed tests for Python 2.7 -- can you take a look please and fix it if possible in a separate PR -- thanks :)
> > https://github.com/apache/airflow/runs/1770939841#step:6:821
> 
> Oh dear....yes, I'll look into this!

Thank You, appreciate it :)

**Comment by madison-ookla** [comment]
Follow up PR here: https://github.com/apache/airflow/pull/13981

**Pull request merged**
